{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "# coding: utf-8\n",
    "# <nbformat>3.0</nbformat>\n",
    "\n",
    "# <markdowncell>\n",
    "# <codecell>\n",
    "\n",
    "\n",
    "# In[1]:\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os \n",
    "from collections import Counter\n",
    "from scipy import spatial\n",
    "import sys\n",
    "import networkx as nx\n",
    "sys.setrecursionlimit(10000)\n",
    "\n",
    "\n",
    "def negative_eQTL_oneSNP(eQTL, distance_threshold):\n",
    "    '''\n",
    "    match the eQTL using chr and distance\n",
    "    '''\n",
    "    eSNP_pos = eQTL['SNP_POS']\n",
    "    echr = eQTL['SNP_CHR']\n",
    "    eDistance = eQTL['DISTANCE']\n",
    "    egene = eQTL['GENE']\n",
    "    \n",
    "    genes_chr = GRCh37_genes[GRCh37_genes['Chromosome/scaffold name'] == str(echr)]\n",
    "    t = genes_chr.apply(lambda x: abs(x['Gene Start (bp)'] - eSNP_pos) < distance_threshold, axis=1)\n",
    "    neg_gene = genes_chr[t].sample(1)\n",
    "    \n",
    "    if sum(t)>1:\n",
    "        while neg_gene.index[0] == egene:\n",
    "            neg_gene = genes_chr[t].sample(1)\n",
    "    else: \n",
    "        print \"No negative gene found\"\n",
    "        \n",
    "    return neg_gene.index[0]\n",
    "    \n",
    "\n",
    "\n",
    "# In[3]:\n",
    "\n",
    "def negative_eQTL_onecell(eQTL, distance_threshold = 1000000):\n",
    "    \n",
    "    eQTL_nega = pd.DataFrame()\n",
    "    \n",
    "    for name, df in eQTL.groupby('SNP_CHR'):\n",
    "        eQTL_pergene = df.loc[df.groupby('GENE')['P-VALUE'].idxmin()]    ## One Lead SNP per gene\n",
    "        eQTL_pergene = eQTL_pergene.reset_index(drop=True)\n",
    "        negative_genes = eQTL_pergene.apply(lambda x: negative_eQTL_oneSNP(x,distance_threshold=distance_threshold),axis=1)\n",
    "        negative_genes = GRCh37_genes.loc[np.array(negative_genes)].reset_index(drop=True)\n",
    "\n",
    "        eQTL_nega_pergene = eQTL_pergene.merge(negative_genes, left_index=True, right_index=True)\n",
    "        eQTL_nega_pergene = eQTL_nega_pergene[['SNP', 'SNP_CHR', 'SNP_POS', 'Associated Gene Name', 'Gene Start (bp)','P-VALUE']]\n",
    "        eQTL_nega_pergene.columns = [['SNP', 'SNP_CHR', 'SNP_POS', 'GENE', 'GENE_START_POS','P-VALUE']]\n",
    "        eQTL_nega_pergene['DISTANCE'] = abs(eQTL_nega_pergene['SNP_POS'] - eQTL_nega_pergene['GENE_START_POS'])\n",
    "        \n",
    "        eQTL_nega = eQTL_nega.append(eQTL_nega_pergene)\n",
    "        \n",
    "    return eQTL_nega\n",
    "\n",
    "\n",
    "# In[4]:\n",
    "\n",
    "def eQTL_in_fragments_one_chr(eQTL_df, contacting_df, SNP_window, promoter_window, \n",
    "                              fragments = ['baitStart', 'baitEnd', 'oeStart', 'oeEnd']):\n",
    "    '''\n",
    "    e-SNP near bait/oe, and promoter of the e-gene near oe/bait\n",
    "    '''\n",
    "    \n",
    "    N = len(contacting_df)\n",
    "    Start1, End1, Start2, End2 = fragments\n",
    "\n",
    "    # e-SNP\n",
    "    SNP_positions = np.reshape(eQTL_df['SNP_POS'],[len(eQTL_df),1])\n",
    "    SNP_tree = spatial.KDTree(SNP_positions)\n",
    "\n",
    "    frag1_start = np.reshape(contacting_df[Start1],[N,1])\n",
    "    SNP_near_frag1Start = SNP_tree.query_ball_point(frag1_start, SNP_window)\n",
    "    frag1_end = np.reshape(contacting_df[End1],[N,1])\n",
    "    SNP_near_frag1End = SNP_tree.query_ball_point(frag1_end, SNP_window)\n",
    "\n",
    "    SNP_near_frag1 = [SNP_near_frag1Start[i] + SNP_near_frag1End[i] for i in xrange(N)]\n",
    "\n",
    "\n",
    "    # e-gene\n",
    "    gene_start = np.reshape(eQTL_df['GENE_START_POS'], [len(eQTL_df),1])\n",
    "    gene_tree = spatial.KDTree(gene_start)\n",
    "\n",
    "    frag2_start = np.reshape(contacting_df[Start2], [N,1])\n",
    "    gene_near_frag2Start = gene_tree.query_ball_point(frag2_start, promoter_window)\n",
    "    frag2_end = np.reshape(contacting_df[End2], [N,1])\n",
    "    gene_near_frag2End = gene_tree.query_ball_point(frag2_end, promoter_window)\n",
    "\n",
    "    gene_near_frag2 = [gene_near_frag2Start[i] + gene_near_frag2End[i] for i in xrange(N)]\n",
    "    \n",
    "    SNPID = np.array(SNP_near_frag1)[np.array([len(np.intersect1d(SNP_near_frag1[t],gene_near_frag2[t]))>0 \n",
    "                                               for t in xrange(len(SNP_near_frag1))])]\n",
    "    geneID = np.array(gene_near_frag2)[np.array([len(np.intersect1d(SNP_near_frag1[t],gene_near_frag2[t]))>0 \n",
    "                                                 for t in xrange(len(SNP_near_frag1))])]\n",
    "    QTLID = [list(np.intersect1d(SNPID[t], geneID[t])) for t in xrange(len(SNPID))]\n",
    "    \n",
    "    return list(set([x for sublist in QTLID for x in sublist]))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# In[5]:\n",
    "\n",
    "def eQTL_in_fragments_one_cell(eQTL, pairs, cell, SNP_window, promoter_window):\n",
    "    \n",
    "    contacting_cell = pairs[list(pairs.columns[:11])+[cell]]  \n",
    "    contacting_cell = contacting_cell[contacting_cell[cell] > 0]   \n",
    "\n",
    "    result_df = pd.DataFrame()\n",
    "    for name, df in eQTL.groupby('SNP_CHR'):\n",
    "        contacting_pairs = contacting_cell[contacting_cell['baitChr'] == str(name)]\n",
    "        eQTL_pergene = df.loc[df.groupby('GENE')['P-VALUE'].idxmin()]    ## One Lead SNP per gene\n",
    "        \n",
    "        functional_contacting_idx1 = eQTL_in_fragments_one_chr(eQTL_pergene, contacting_pairs, SNP_window, promoter_window)\n",
    "        functional_contacting_idx2 = eQTL_in_fragments_one_chr(eQTL_pergene, contacting_pairs, SNP_window, promoter_window,\n",
    "                                                               fragments = ['oeStart', 'oeEnd','baitStart', 'baitEnd'])\n",
    "        contact_vector = np.array([0] * len(eQTL_pergene))\n",
    "        contact_vector[np.array(list(set(functional_contacting_idx1+functional_contacting_idx2)))] = 1\n",
    "        eQTL_pergene['contacting'] = contact_vector\n",
    "        \n",
    "        result_df = result_df.append(eQTL_pergene)\n",
    "        \n",
    "        print 'There are %i contacting eQTL pairs among %i for chr%i' % (sum(contact_vector),len(eQTL_pergene), name)\n",
    "                \n",
    "    return result_df\n",
    "\n",
    "\n",
    "\n",
    "# In[6]:\n",
    "\n",
    "def save_bed_files(result_df, celltype, SNP_window, promoter_window, negative_set = False):\n",
    "    '''\n",
    "    save the bed files for intersection with TFB motif\n",
    "    '''\n",
    "    \n",
    "    SNP_bed = []\n",
    "    SNP_bed.append(['chr%i'% x for x in (result_df['SNP_CHR'])])\n",
    "    SNP_bed.append(list(np.array(result_df['SNP_POS']) - SNP_window))\n",
    "    SNP_bed.append(list(np.array(result_df['SNP_POS']) + SNP_window))\n",
    "    SNP_bed.append(list(result_df['SNP']))\n",
    "    SNP_bed = pd.DataFrame(SNP_bed).transpose()\n",
    "    if negative_set:\n",
    "        promoter_bed.to_csv(os.path.join(DATA_DIR, 'intermediate/%s/PCHiC_peak_matrix_cutoff5_%s_eSNP_negativeset.bed'% (celltype, celltype)),\n",
    "                   sep='\\t',index=False, header=False)\n",
    "    else:\n",
    "        SNP_bed.to_csv(os.path.join(DATA_DIR, 'intermediate/%s/PCHiC_peak_matrix_cutoff5_%s_eSNP.bed' % (celltype, celltype)),\n",
    "                   sep='\\t',index=False, header=False)\n",
    "    \n",
    "    ## \n",
    "    promoter_bed = []\n",
    "    promoter_bed.append(['chr%i'% x for x in (result_df['SNP_CHR'])])\n",
    "    promoter_bed.append(list(np.array(result_df['GENE_START_POS']) - promoter_window))\n",
    "    promoter_bed.append(list(np.array(result_df['GENE_START_POS']) + promoter_window))\n",
    "    promoter_bed.append(list(result_df['GENE']))\n",
    "    promoter_bed = pd.DataFrame(promoter_bed).transpose()\n",
    "    if negative_set:\n",
    "        promoter_bed.to_csv(os.path.join(DATA_DIR, 'intermediate/%s/PCHiC_peak_matrix_cutoff5_%s_egene_negativeset.bed'% (celltype, celltype)),\n",
    "                   sep='\\t',index=False, header=False)\n",
    "    else:\n",
    "        promoter_bed.to_csv(os.path.join(DATA_DIR, 'intermediate/%s/PCHiC_peak_matrix_cutoff5_%s_egene.bed'% (celltype, celltype)),\n",
    "                   sep='\\t',index=False, header=False)\n",
    "        \n",
    "        \n",
    "def readin_intermediate_result(celltype, SNP_window, promoter_window, negative_set=False):\n",
    "    if negative_set:\n",
    "        fn_SNP = 'PCHiC_peak_matrix_cutoff5_%s_eSNP_negative.bed' % chr\n",
    "        fn_gene = 'PCHiC_peak_matrix_cutoff5_%s_egene_negative.bed' % chr\n",
    "    else:\n",
    "        fn_SNP = 'PCHiC_peak_matrix_cutoff5_%s_eSNP.bed' % chr\n",
    "        fn_gene = 'PCHiC_peak_matrix_cutoff5_%s_egene.bed' % chr\n",
    "    a = pd.read_csv(os.path.join(DATA_DIR, 'intermediate/%s/%s' % (celltype,fn_SNP)),sep='\\t',header=None)\n",
    "    b = pd.read_csv(os.path.join(DATA_DIR, 'intermediate/%s/%s' % (celltype,fn_gene)),sep='\\t',header=None)\n",
    "    \n",
    "    inter_result = pd.concat((a,b),axis=1)\n",
    "    inter_result.columns = ['SNP_CHR','SNP_POS','non','SNP','GENE_chr','GENE_START_POS','NAN','GENE']\n",
    "    inter_result = inter_result[['SNP_CHR','SNP_POS','SNP','GENE_chr','GENE_START_POS','GENE']]\n",
    "    inter_result['SNP_POS'] = inter_result['SNP_POS'] + SNP_window\n",
    "    inter_result['GENE_START_POS'] = inter_result['GENE_START_POS'] + promoter_window\n",
    "    return inter_result\n",
    "\n",
    "# In[ ]:\n",
    "\n",
    "#### Need to scan the bed files for known motifs using fimo\n",
    "#### run the following codes in DATA_DIR\n",
    "#### bash TFBS.sh Mon eSNP\n",
    "#### bash TFBS.sh Mon gene\n",
    "#### bash TFBS.sh tCD4 eSNP\n",
    "#### bash TFBS.sh tCD4 gene\n",
    "\n",
    "\n",
    "# In[ ]:\n",
    "\n",
    "-\n",
    "\n",
    "# In[78]:\n",
    "\n",
    "def eQTL_in_ATACseq_one_chr(eQTL_pos, ATAC_regions, window = 1000):\n",
    "\n",
    "    N = len(ATAC_regions)\n",
    "    eQTL_tree = spatial.KDTree(eQTL_pos)\n",
    "\n",
    "    frag_start = np.reshape(ATAC_regions['Start'],[N,1])\n",
    "    eQTL_near_fragStart = eQTL_tree.query_ball_point(frag_start, window)\n",
    "    frag_end = np.reshape(ATAC_regions['End'],[N,1])\n",
    "    eQTL_near_fragEnd = eQTL_tree.query_ball_point(frag_end, window)\n",
    "\n",
    "    eQTL_near_frag = [list(set(eQTL_near_fragStart[i] + eQTL_near_fragEnd[i])) for i in xrange(N)]\n",
    "    eQTLID = [x for x in eQTL_near_frag if len(x) > 0]\n",
    "    eQTLID = list(set([a for b in eQTLID for a in b]))\n",
    "    \n",
    "    x = np.zeros(len(eQTL_pos))\n",
    "    x[eQTLID] = 1\n",
    "    return x\n",
    "\n",
    "\n",
    "# In[119]:\n",
    "\n",
    "def eQTL_in_ATACseq_one_cell(eQTL, ATAC, window_SNP = 1000, window_gene = 2000000):\n",
    "    \n",
    "    result_df = pd.DataFrame()\n",
    "    for name, g in eQTL.groupby('SNP_CHR'):\n",
    "        ## eSNP\n",
    "        eSNP_pos = np.reshape(g['SNP_POS'],[len(g),1])\n",
    "        ATAC_SNP_vector = eQTL_in_ATACseq_one_chr(eSNP_pos,ATAC[ATAC['Chr'] == name], window = window_SNP)\n",
    "        ## egene\n",
    "        egene_pos = np.reshape(g['GENE_START_POS'],[len(g),1])\n",
    "        ATAC_gene_vector = eQTL_in_ATACseq_one_chr(egene_pos,ATAC[ATAC['Chr'] == name], window = window_gene)\n",
    "        \n",
    "        ## ATAC profile\n",
    "        ATAC_eQTL = pd.DataFrame({'ATAC_SNP':ATAC_SNP_vector,'ATAC_gene':ATAC_gene_vector})\n",
    "        g = pd.concat((g.reset_index(drop=True),ATAC_eQTL),axis=1)\n",
    "        result_df = result_df.append(g)        \n",
    "        \n",
    "    return result_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def SNP_in_PIR(eQTL, contacting_df, PIR_window = 500):\n",
    "    gene_promoter = eQTL['GENE']\n",
    "    SNP = eQTL['SNP']\n",
    "    SNP_POS = eQTL['SNP_POS']\n",
    "\n",
    "    gene_promoter_idx = np.where([gene_promoter in x.split(';') for x in list(contacting_df['baitName'])])[0]\n",
    "    gene_promoter_window = contacting_df.iloc[gene_promoter_idx]\n",
    "\n",
    "    left = np.array(gene_promoter_window['oeStart']) - PIR_window\n",
    "    right = np.array(gene_promoter_window['oeEnd']) + PIR_window\n",
    "\n",
    "    if sum([(t[0] < SNP_POS and t[1] > SNP_POS) for t in zip(left,right)]) > 0:\n",
    "        return 1\n",
    "    else:\n",
    "        return 0\n",
    "    \n",
    "    \n",
    "    \n",
    "def eQTL_in_fragments_one_cell(eQTL, pairs, cell, SNP_window, promoter_window):\n",
    "    \n",
    "    contacting_cell = pairs[list(pairs.columns[:11])+[cell]]  \n",
    "    contacting_cell = contacting_cell[contacting_cell[cell] > 0]   \n",
    "\n",
    "    result_df = pd.DataFrame()\n",
    "    for name, df in eQTL.groupby('SNP_CHR'):\n",
    "        contacting_pairs = contacting_cell[contacting_cell['baitChr'] == str(name)]\n",
    "        eQTL_pergene = df.loc[df.groupby('GENE')['P-VALUE'].idxmin()]    ## One Lead SNP per gene\n",
    "        \n",
    "        functional_contacting_idx = eQTL_pergene.apply(lambda x: SNP_in_PIR(x, contacting_pairs),axis=1)\n",
    "        contact_vector = np.array(functional_contacting_idx)\n",
    "        eQTL_pergene['contacting'] = contact_vector\n",
    "        \n",
    "        result_df = result_df.append(eQTL_pergene)\n",
    "        \n",
    "        print 'There are %i contacting eQTL pairs among %i for chr%i' % (sum(contact_vector),len(eQTL_pergene), name)\n",
    "                \n",
    "    return result_df\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DATA_DIR = '/Users/Yuan/Documents/BLab/Predict_target_genes/data'\n",
    "\n",
    "compute_negative_set = True\n",
    "match_distance = 1000000\n",
    "\n",
    "from_beginning=True\n",
    "negative_set_flag = True\n",
    "\n",
    "SNP_window_for_PHiC = 1000\n",
    "promoter_window_for_PHiC = 1000000\n",
    "\n",
    "SNP_window_for_motif = 1000\n",
    "promoter_window_for_motif = 2000\n",
    "\n",
    "if compute_negative_set:\n",
    "    ### read in all genes infomation\n",
    "    GRCh37_genes = pd.read_csv(os.path.join(DATA_DIR, 'eQTL/GRCh37_genes.txt'),sep='\\t')\n",
    "    gene_idx = []\n",
    "    for name, df in GRCh37_genes.groupby(['Associated Gene Name']):\n",
    "        gene_idx.append(np.random.choice(list(df.index)))\n",
    "    GRCh37_genes = GRCh37_genes.iloc[np.array(gene_idx)]\n",
    "    GRCh37_genes.index = GRCh37_genes['Associated Gene Name']\n",
    "    \n",
    "    ### compute negative sets match for chr and distance\n",
    "    CD4_eQTL = pd.read_csv(os.path.join(DATA_DIR, 'eQTL/Raj/cd4T_cis_fdr05.tsv'),sep='\\t')\n",
    "    mon_eQTL = pd.read_csv(os.path.join(DATA_DIR, 'eQTL/Raj/monocytes_cis_fdr05.tsv'),sep='\\t')\n",
    "    CD4_eQTL_neg = negative_eQTL_onecell(CD4_eQTL, match_distance)\n",
    "    CD4_eQTL_neg.to_csv(os.path.join(DATA_DIR, 'eQTL/Raj/cd4T_cis_fdr05_negative.tsv'),sep='\\t',index=False)\n",
    "    mon_eQTL_neg = negative_eQTL_onecell(mon_eQTL, match_distance)\n",
    "    mon_eQTL_neg.to_csv(os.path.join(DATA_DIR, 'eQTL/Raj/monocytes_cis_fdr05_negative.tsv'),sep='\\t',index=False)\n",
    "\n",
    "    \n",
    "if from_beginning:\n",
    "    ### First use all contacting pairs (could transfer to contacting E_P pairs later)\n",
    "    pairs = pd.read_csv(os.path.join(DATA_DIR,'CPHiC/Interactions/PCHiC_peak_matrix_cutoff5.txt'),sep='\\t', low_memory=False)\n",
    "    pairs = pairs[pairs['baitChr'] == pairs['oeChr']]\n",
    "    \n",
    "    ### read in the (negative) eQTL list\n",
    "    if negative_set_flag:\n",
    "        CD4_eQTL = pd.read_csv(os.path.join(DATA_DIR, 'eQTL/Raj/cd4T_cis_fdr05_negative.tsv'),sep='\\t')\n",
    "        mon_eQTL = pd.read_csv(os.path.join(DATA_DIR, 'eQTL/Raj/monocytes_cis_fdr05_negative.tsv'),sep='\\t')\n",
    "    else:\n",
    "        CD4_eQTL = pd.read_csv(os.path.join(DATA_DIR, 'eQTL/Raj/cd4T_cis_fdr05.tsv'),sep='\\t')\n",
    "        mon_eQTL = pd.read_csv(os.path.join(DATA_DIR, 'eQTL/Raj/monocytes_cis_fdr05.tsv'),sep='\\t')\n",
    "        \n",
    "    ### compute the overlap of baits/oe with the eQTL list\n",
    "    print \"tCD4\"\n",
    "    tCD4_result = eQTL_in_fragments_one_cell(CD4_eQTL, pairs, 'tCD4',SNP_window_for_PHiC, promoter_window_for_PHiC)\n",
    "    save_bed_files(tCD4_result, 'tCD4', SNP_window_for_motif, promoter_window_for_motif, negative_set_flag)\n",
    "    print \"Mon\"\n",
    "    Mon_result = eQTL_in_fragments_one_cell(mon_eQTL, pairs, 'Mon',SNP_window_for_PHiC, promoter_window_for_PHiC)\n",
    "    save_bed_files(Mon_result, 'Mon', SNP_window_for_motif, promoter_window_for_motif, negative_set_flag)\n",
    "    \n",
    "else:\n",
    "    ### read directly from the intermediate results\n",
    "    if negative_set_flag:\n",
    "        Mon_result = pd.read_csv(os.path.join(DATA_DIR, 'eQTL/FairFax/%s_cis_%i_fdr05_50kb_negative_annotated.csv' % (celltype,chr)), sep='\\t')\n",
    "    else:\n",
    "        Mon_result = pd.read_csv(os.path.join(DATA_DIR, 'eQTL/FairFax/%s_cis_%i_fdr05_50kb_annotated.csv' % (celltype,chr)), sep='\\t')\n",
    "        \n",
    "    ### add on the TF motif information\n",
    "    Mon_result = merge_TF_motif(Mon_result, 'Mon')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>SNP_CHR</th>\n",
       "      <th>SNP_POS</th>\n",
       "      <th>GENE</th>\n",
       "      <th>GENE_START_POS</th>\n",
       "      <th>P-VALUE</th>\n",
       "      <th>DISTANCE</th>\n",
       "      <th>SNP_contacting_df_idx</th>\n",
       "      <th>contacting</th>\n",
       "      <th>motif_SNP</th>\n",
       "      <th>motif_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs6701264</td>\n",
       "      <td>1</td>\n",
       "      <td>34272479</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>33545037</td>\n",
       "      <td>2.120000e-04</td>\n",
       "      <td>727442</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>[MA0050.2]</td>\n",
       "      <td>[MA0014.2, MA0050.2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs11804707</td>\n",
       "      <td>1</td>\n",
       "      <td>10689053</td>\n",
       "      <td>AADACL4</td>\n",
       "      <td>12649347</td>\n",
       "      <td>1.602000e-03</td>\n",
       "      <td>1960294</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>[MA0050.2]</td>\n",
       "      <td>[MA0050.2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs7518038</td>\n",
       "      <td>1</td>\n",
       "      <td>75927018</td>\n",
       "      <td>ACADM</td>\n",
       "      <td>75999445</td>\n",
       "      <td>4.940000e-07</td>\n",
       "      <td>72427</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>[MA0050.2]</td>\n",
       "      <td>[MA0050.2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs2986739</td>\n",
       "      <td>1</td>\n",
       "      <td>6470643</td>\n",
       "      <td>ACOT7</td>\n",
       "      <td>6246931</td>\n",
       "      <td>1.260000e-05</td>\n",
       "      <td>223712</td>\n",
       "      <td>[818, 841, 854, 971, 1014]</td>\n",
       "      <td>2</td>\n",
       "      <td>[MA0050.2]</td>\n",
       "      <td>[MA0014.2, MA0139.1, MA0050.2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs12040970</td>\n",
       "      <td>1</td>\n",
       "      <td>153361924</td>\n",
       "      <td>ADAM15</td>\n",
       "      <td>153299393</td>\n",
       "      <td>5.120000e-09</td>\n",
       "      <td>62531</td>\n",
       "      <td>[36135]</td>\n",
       "      <td>0</td>\n",
       "      <td>[MA0050.2]</td>\n",
       "      <td>[MA0050.2]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          SNP  SNP_CHR    SNP_POS     GENE  GENE_START_POS       P-VALUE  \\\n",
       "0   rs6701264        1   34272479  A3GALT2        33545037  2.120000e-04   \n",
       "1  rs11804707        1   10689053  AADACL4        12649347  1.602000e-03   \n",
       "2   rs7518038        1   75927018    ACADM        75999445  4.940000e-07   \n",
       "3   rs2986739        1    6470643    ACOT7         6246931  1.260000e-05   \n",
       "4  rs12040970        1  153361924   ADAM15       153299393  5.120000e-09   \n",
       "\n",
       "   DISTANCE       SNP_contacting_df_idx  contacting   motif_SNP  \\\n",
       "0    727442                          []           0  [MA0050.2]   \n",
       "1   1960294                          []           0  [MA0050.2]   \n",
       "2     72427                          []           0  [MA0050.2]   \n",
       "3    223712  [818, 841, 854, 971, 1014]           2  [MA0050.2]   \n",
       "4     62531                     [36135]           0  [MA0050.2]   \n",
       "\n",
       "                       motif_gene  \n",
       "0            [MA0014.2, MA0050.2]  \n",
       "1                      [MA0050.2]  \n",
       "2                      [MA0050.2]  \n",
       "3  [MA0014.2, MA0139.1, MA0050.2]  \n",
       "4                      [MA0050.2]  "
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# In[120]:\n",
    "\n",
    "#### ATAC-seq\n",
    "\n",
    "ATACseq = pd.read_csv(os.path.join(DATA_DIR, 'ATACseq/GSE74912_ATACseq_All_Counts.txt'),sep='\\t')\n",
    "ATAC_CD4 = pd.concat((ATACseq[['Chr','Start','End']],\n",
    "#                      ATACseq[[x for x in ATACseq.columns if 'CD4' in x]].mean(axis=1)),axis=1)\n",
    "tCD4_TF_ATAC = eQTL_in_ATACseq_one_cell(tCD4_TF, ATACseq)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def map_to_1based(anarray):\n",
    "    return dict(zip(set(anarray), xrange(len(anarray))))\n",
    "\n",
    "\n",
    "def bait_names(mapped_1based_idx, names):\n",
    "    return dict(zip(mapped_1based_idx, names))\n",
    "\n",
    "def edges_inbetween(D, n_nodes):\n",
    "    ### input: a graph object\n",
    "    ### return the degree of contacts between nodes: 1/2/0\n",
    "    goal_matrix = np.ones([n_nodes,n_nodes]) * 100\n",
    "    for n,nbrs in D.adjacency_iter():\n",
    "        # assign 1 to the direct linkage\n",
    "        goal_matrix[n][D[n].keys()] = 1\n",
    "        # assign 2 to the indirect linkage\n",
    "        for t in D[n].keys():\n",
    "            goal_matrix[n][D[t].keys()] = map(lambda x: min(x), \n",
    "                                              zip(list(goal_matrix[n][D[t].keys()]),\n",
    "                                                  [goal_matrix[n][t]+1] * len(goal_matrix[n][D[t].keys()])))\n",
    "    goal_matrix[goal_matrix==100] = 0\n",
    "    np.fill_diagonal(goal_matrix,0)\n",
    "    return goal_matrix\n",
    "\n",
    "\n",
    "\n",
    "def contacting_to_degree(data,edge_threshold=0):\n",
    "    ## obtain the nodes and the weights edges\n",
    "    if edge_threshold != 0:\n",
    "        data = data.iloc[np.where(data['obs_exp']>edge_threshold)[0]]\n",
    "    fragment_len = max(max(data['i']),max(data['j'])) + 1\n",
    "    nodes = range(fragment_len)\n",
    "    weighted_edges = zip(list(data['i']),list(data['j']),list(data['obs_exp']))\n",
    "    print fragment_len\n",
    "    \n",
    "    ## construct the graph\n",
    "    G = nx.Graph()\n",
    "    G.add_nodes_from(nodes)\n",
    "    G.add_weighted_edges_from(weighted_edges)\n",
    "    \n",
    "    ## obtain number of edges between the bins\n",
    "    number_of_edges_inbetween = edges_inbetween(G, fragment_len)\n",
    "    return G, number_of_edges_inbetween"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def compute_frag_by_frag(d, celltype):\n",
    "    anarray = list(d['baitID']) + list(d['oeID'])\n",
    "    t = map_to_1based(anarray)\n",
    "    d['baitID'] = map(lambda x: t[x], list(d['baitID']))\n",
    "    d['oeID'] = map(lambda x: t[x], list(d['oeID']))\n",
    "    baitsnames = bait_names(list(d['baitID']) + list(d['oeID']), list(d['baitName']) + list(d['oeName']))\n",
    "\n",
    "    df = d[['baitID','oeID',celltype]]\n",
    "    df.columns = ['i','j','obs_exp']\n",
    "\n",
    "    df = df.reset_index(drop=True)\n",
    "\n",
    "    G,frag_by_frag = contacting_to_degree(df,edge_threshold=0)\n",
    "    frag_by_frag = pd.DataFrame(frag_by_frag)\n",
    "    frag_by_frag.index = baitsnames.values()\n",
    "    frag_by_frag.loc['NOGENE'] = [0]*len(frag_by_frag)\n",
    "    \n",
    "    frag_by_frag = frag_by_frag.drop([\".\"])\n",
    "    frag_by_frag = frag_by_frag.to_dict(orient = 'index')\n",
    "    \n",
    "    return frag_by_frag"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def find_the_key(frag_by_frag, gg):\n",
    "    gg_in_frag = np.where(map(lambda x: gg in x, frag_by_frag.keys()))[0]\n",
    "    if len(gg_in_frag) > 0:\n",
    "        temp = np.array(frag_by_frag.keys())[np.array(gg_in_frag)]\n",
    "        return np.random.choice(temp)\n",
    "    else:\n",
    "        return 'NOGENE'\n",
    "    \n",
    "\n",
    "def degree_of_contact_SNP_gene(eQTL_df, contacting_df, frag_by_frag):\n",
    "\n",
    "    ### find the other fragments where eSNP locate in \n",
    "    left = np.array(contacting_df['oeStart']) - 500\n",
    "    right = np.array(contacting_df['oeEnd']) + 500\n",
    "    SNPs_PIR = eQTL_df.apply(lambda x: [x['SNP_POS'] in range(t[0],t[1]) for t in zip(left,right)], axis=1)\n",
    "    SNPs_PIR = SNPs_PIR.apply(lambda x: list(np.where(x)[0]))   ## a list of lists\n",
    "    eQTL_df['SNP_contacting_df_idx'] = list(SNPs_PIR)\n",
    "    \n",
    "    ### find the degree of contact by using frag_by_frag matrix\n",
    "    degree_contact = eQTL_df.apply(lambda eq: [frag_by_frag[find_the_key(frag_by_frag, eq['GENE'])][contacting_df.iloc[t]['oeID']] \n",
    "                                               for t in eq['SNP_contacting_df_idx']], axis=1)\n",
    "    \n",
    "    return degree_contact"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DATA_DIR = '/Users/Yuan/Documents/BLab/Predict_target_genes/data'\n",
    "\n",
    "if 1:\n",
    "    pairs = pd.read_csv(os.path.join(DATA_DIR,'CPHiC/Interactions/PCHiC_peak_matrix_cutoff5.txt'),sep='\\t', low_memory=False)\n",
    "    pairs = pairs[pairs['baitChr'] == pairs['oeChr']]\n",
    "    \n",
    "    CD4_eQTL = pd.read_csv(os.path.join(DATA_DIR, 'eQTL/Raj/cd4T_cis_fdr05_negative.tsv'),sep='\\t')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contacting_df = pairs[pairs['baitChr'] == '22']\n",
    "eQTL_df = CD4_eQTL[CD4_eQTL['SNP_CHR'] == 22]\n",
    "eQTL_df = eQTL_df.iloc[10:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2904\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Python/2.7/site-packages/ipykernel/__main__.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/Library/Python/2.7/site-packages/ipykernel/__main__.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "frag2 = compute_frag_by_frag(contacting_df,'tCD4') \n",
    "degree_contact = degree_of_contact_SNP_gene(eQTL_df, contacting_df, frag2)\n",
    "\n",
    "contact_vector = np.zeros(len(eQTL_df))\n",
    "contact_vector[np.where(map(lambda x: 1 in x, degree_contact))[0]] = 1\n",
    "contact_vector[np.where(map(lambda x: 2 in x, degree_contact))[0]] = 2\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compute whether the eQTL pair fall in any promoter-PIR pair\n",
      "2817\n",
      "22\n",
      "Counter({0.0: 9, 1.0: 1})\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Python/2.7/site-packages/ipykernel/__main__.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/Library/Python/2.7/site-packages/ipykernel/__main__.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "if 1:\n",
    "    eQTL = eQTL_df\n",
    "    cell = 'tCD4'\n",
    "    \n",
    "    print 'Compute whether the eQTL pair fall in any promoter-PIR pair'\n",
    "    \n",
    "    contacting_cell = pairs[list(pairs.columns[:11])+[cell]]  \n",
    "    contacting_cell = contacting_cell[contacting_cell[cell] > 0]   \n",
    "\n",
    "    result_df = pd.DataFrame()\n",
    "    for name, df in eQTL.groupby('SNP_CHR'):\n",
    "        contacting_pairs = contacting_cell[contacting_cell['baitChr'] == str(name)]\n",
    "        eQTL_pergene = df.loc[df.groupby('GENE')['P-VALUE'].idxmin()]    ## One Lead SNP per gene\n",
    "        \n",
    "        # functional_contacting_idx1 = eQTL_in_fragments_one_chr(eQTL_pergene, contacting_pairs, SNP_window, promoter_window)\n",
    "        # functional_contacting_idx2 = eQTL_in_fragments_one_chr(eQTL_pergene, contacting_pairs, SNP_window, promoter_window,\n",
    "        #                                                        fragments = ['oeStart', 'oeEnd','baitStart', 'baitEnd'])\n",
    "        # contact_vector = np.array([0] * len(eQTL_pergene))\n",
    "        # contact_vector[np.array(list(set(functional_contacting_idx1+functional_contacting_idx2)))] = 1\n",
    "\n",
    "        frag2 = compute_frag_by_frag(contacting_pairs,cell) \n",
    "        degree_contact = degree_of_contact_SNP_gene(eQTL_pergene, contacting_pairs, frag2)\n",
    "\n",
    "        contact_vector = np.zeros(len(eQTL_pergene))\n",
    "        contact_vector[np.where(map(lambda x: 1 in x, degree_contact))[0]] = 1\n",
    "        contact_vector[np.where(map(lambda x: 2 in x, degree_contact))[0]] = 2\n",
    "\n",
    "        eQTL_pergene['contacting'] = contact_vector\n",
    "        \n",
    "        result_df = result_df.append(eQTL_pergene)\n",
    "        print name\n",
    "        print Counter(contact_vector)\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n",
      "False\n"
     ]
    }
   ],
   "source": [
    "temp = [x for sublist in [str(x).split(';') for x in pairs['baitName']] for x in sublist]\n",
    "print t in temp\n",
    "\n",
    "temp = [x for sublist in [str(x).split(';') for x in pairs['oeName']] for x in sublist]\n",
    "print t in temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "26643"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = [x for sublist in [str(x).split(';') for x in pairs['baitName']] for x in sublist]\n",
    "len(list(set(temp)))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
